src/main/scala/mesosphere/marathon/core/matcher/reconcile/impl/OffersWantedForReconciliationActor.scala

```diff
@@ -77,42 +77,43 @@ private[reconcile] class OffersWantedForReconciliationActor(
       }
   }
 
-  private[this] def switchToSubscribedToOffers(reason: String): Receive = {
-    val nextCheck = scheduleNextCheck
+  private[this] def subscribedToOffers(reason: String): Receive = {
+    // exit this state after the timeout
+    val timeout = scheduleCancelInterestInOffers
```

Comment:
I think that there is a race condition here. I am pretty certain that calling timeout.cancel() down below will not result in a guarantee that this cancellation is not sent.

To resolve, perhaps we can change `CancelInterestInOffers` to a case class with the deadline that is being cancelled.

```
    val (deadline, timeout) = scheduleCancelInterestInOffers
    ...
    LoggingReceive.withLabel("subscribedToOffers") {
      handleRequestOfferIndicators orElse {
        case OffersWantedForReconciliationActor.CancelInterestInOffers(cancelDeadline) =>
          if (cancelDeadline >= deadline) {
            ...
          } else { /* ignore .... */ }
```

```
  protected def scheduleCancelInterestInOffers: (Instant, Cancellable) = {
    val deadline = clock.now() + interestDuration
    val timeout = context.system.scheduler.scheduleOnce(
      interestDuration, self, OffersWantedForReconciliationActor.CancelInterestInOffers(deadline)
    )(context.dispatcher)
    (deadline, timeout)
  }
```

This way, if an old deadline cancellation signal is received, it is simply ignored, since the timestamp is in the past

Reply:
If I'm not mistaken that's the logic we had before. Isn't the core of the algorithm:

1. I'm interested in offers until X = now+duration
2. a) Reached X -> I'm not interested anymore b) I received a new interest thus I override X again.

So we just need to check if the current X, aka deadline is reached and ignore older deadlines.... which is what you proposed I guess. Maybe we should use UUIDs instead

```scala
val currentCancelId = scheduleCancelInterestInOffers
    LoggingReceive.withLabel("subscribedToOffers") {
      handleRequestOfferIndicators orElse {
        case OffersWantedForReconciliationActor.CancelInterestInOffers(cancelId) =>
          if (currentCancelId = cancelId) {
            ...
          } else { /* ignore .... */ }
...

  protected def scheduleCancelInterestInOffers: UUID = {
    val  id = new UUID()
    val timeout = context.system.scheduler.scheduleOnce(
      interestDuration, self, OffersWantedForReconciliationActor.CancelInterestInOffers(id)
    )(context.dispatcher)
    id
  }

```

Reply:
I liked @jeschkies 's approach slightly more :)